File: fsl-blog-scala/src/main/scala/futures/ProducerConsumer.scala
package futures

import org.apache.commons.lang3.RandomStringUtils

import java.util.concurrent.{BlockingQueue, Executors, LinkedBlockingQueue}
import scala.annotation.tailrec
import scala.concurrent.ExecutionContext
import scala.util.Random
import scala.concurrent.Future

object ProducerConsumer {

  implicit val executionContext = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(20))

  def getRandomTime: Int = Random.between(500, 1001)

  def producer(id: Long, queue: BlockingQueue[String]): Future[Unit] = Future {
    produce(id, queue)
  }

  @tailrec
  def produce(id: Long, queue: BlockingQueue[String]): Unit = {
    Thread.sleep(getRandomTime)

    val value = RandomStringUtils.randomAlphanumeric(10)
    queue.put(value)
    println("producer " + id + " producing " + value)

    produce(id, queue)
  }

  def consumer(id: Long, queue: BlockingQueue[String]): Future[Unit] = Future {
    consume(id, queue)
  }

  @tailrec
  def consume(id: Long, queue: BlockingQueue[String]): Unit = {
    Thread.sleep(getRandomTime)

    val value = queue.take()
    println("consumer " + id + " consuming value " + value)

    consume(id, queue)
  }

  def main(args: Array[String]): Unit = {
    val queue =  new LinkedBlockingQueue[String]()

    List.range(1, 11).foreach(producer(_, queue))

    List.range(1, 11).foreach(consumer(_, queue))
  }

}

File: fsl-blog-java/src/main/java/virtualthreads/ProducerConsumer.java
package virtualthreads;

import org.apache.commons.lang3.RandomStringUtils;

import java.util.Random;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.stream.IntStream;

public class ProducerConsumer {
    static int getRandomTime(Random random) {
        return random.ints(500, 1000).findFirst().getAsInt();
    }

    public static void main(String[] args) throws InterruptedException {
        BlockingQueue<String> queue = new LinkedBlockingQueue<>();
        Random random = new Random();

        IntStream.range(1, 11).forEach(id -> {
            Runnable producer = (() -> {
                while (true) {
                    try {
                        Thread.sleep(getRandomTime(random));

                        var value = RandomStringUtils.randomAlphanumeric(10);
                        queue.put(value);
                        System.out.println("producer " + id + " producing " + value);
                    } catch (InterruptedException e) {
                        System.out.println("producer " + id + " error producing value " + e.getMessage());
                    }
                }
            });

            var task = Thread.ofVirtual().unstarted(producer);
            task.start();
        });

        IntStream.range(1, 11).forEach(id -> {
            Runnable consumer = (() -> {
                while (true) {
                    try {
                        Thread.sleep(getRandomTime(random));

                        var value = queue.take();
                        System.out.println("consumer " + id + " consuming value " + value);
                    } catch (InterruptedException e) {
                        System.out.println("consumer " + id + " error consuming value " + e.getMessage());
                    }
                }
            });

            var task = Thread.ofVirtual().unstarted(consumer);
            task.start();
        });

        Thread.sleep(30000);
    }
}
